File: src/clang/magma/compile/Node.c
#include "Node.h"
struct Node withString(struct String propertyKey, struct String propertyValue);
struct Option_String findString(struct String propertyKey);
struct Node withNodeList(struct String propertyKey, struct List__Node propertyValues);
struct Option_List__Node findNodeList(struct String propertyKey);
struct String display();
struct String format(int depth);
struct Node mapNodeList(struct String propertyKey, struct List__Node(*mapper)(struct List__Node));
int is(struct String type);
struct Node retype(struct String type);
struct Node merge(struct Node other);
struct Stream_Tuple_String_String streamStrings();
struct Stream_Tuple_String_List__Node streamNodeLists();
struct Node withNode(struct String propertyKey, struct Node propertyValue);
struct Option_Node findNode(struct String propertyKey);
struct Stream_Tuple_String_Node streamNodes();
struct Node mapNode(struct String propertyKey, struct Node(*mapper)(struct Node));
struct Node withNodeLists(struct Map__String_List__Node nodeLists);
struct Node withNodes(struct Map__String_Node nodes);
struct Node removeNode(struct String propertyKey);
int hasNode(struct String propertyKey);
int hasNodeList(struct String propertyKey);

File: src/java/magma/compile/Node.java
package magma.compile;

import magma.collect.list.List_;
import magma.collect.map.Map_;
import magma.collect.stream.Stream;
import magma.option.Option;
import magma.option.Tuple;

import java.util.function.Function;

public interface Node {
    Node withString(String propertyKey, String propertyValue);

    Option<String> findString(String propertyKey);

    Node withNodeList(String propertyKey, List_<Node> propertyValues);

    Option<List_<Node>> findNodeList(String propertyKey);

    String display();

    String format(int depth);

    Node mapNodeList(String propertyKey, Function<List_<Node>, List_<Node>> mapper);

    boolean is(String type);

    Node retype(String type);

    Node merge(Node other);

    Stream<Tuple<String, String>> streamStrings();

    Stream<Tuple<String, List_<Node>>> streamNodeLists();

    Node withNode(String propertyKey, Node propertyValue);

    Option<Node> findNode(String propertyKey);

    Stream<Tuple<String, Node>> streamNodes();

    Node mapNode(String propertyKey, Function<Node, Node> mapper);

    Node withNodeLists(Map_<String, List_<Node>> nodeLists);

    Node withNodes(Map_<String, Node> nodes);

    Node removeNode(String propertyKey);

    boolean hasNode(String propertyKey);

    boolean hasNodeList(String propertyKey);

    Boolean equalsTo(Node other);

    boolean hasNoType();
}
